DOCSP-40792: Delete guide (#33)
mongodb/docs-cpp: source/write/delete.txt

   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: remove, drop, code example

Overview
--------

In this guide, you can learn how to use the {+driver-short+} to remove
documents from a MongoDB collection by performing **delete operations**.

A delete operation removes one or more documents from a MongoDB collection.
You can perform a delete operation by using the ``delete_one()`` or
``delete_many()`` methods.

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``restaurants`` collection in the ``sample_restaurants``
database from the :atlas:`Atlas sample datasets </sample-data>`. To access this collection
from your C++ application, instantiate a ``mongocxx::client`` that connects to an Atlas cluster
and assign the following values to your ``db`` and ``collection`` variables:

.. literalinclude:: /includes/write/delete.cpp
    :start-after: start-db-coll
    :end-before: end-db-coll
    :language: cpp
    :dedent:

To learn how to create a free MongoDB Atlas cluster and load the sample datasets, see the
:atlas:`Get Started with Atlas </getting-started>` guide.

Delete Operations
-----------------

> You can perform delete operations in MongoDB by using the following methods:

Reviewer: ```suggestion
You can perform delete operations with the {+driver-short+} by using the following methods:
```

- ``delete_one()``, which deletes *the first document* that matches the search criteria
- ``delete_many()``, which deletes *all documents* that match the search criteria

> Each delete method requires a **query filter** document, which specifies the
> search criteria that determine which documents to select for removal.

Reviewer: S:
```suggestion
Each delete method requires a **query filter** document, which specifies the
search criteria that is used to determine which documents to select for removal.
```

         results. For more information, see :manual:`Collation </reference/collation/#std-label-collation>`
         in the {+mdb-server+} manual.

   * - ``write_concern``
     - | Sets the write concern for the operation.
         For more information, see :manual:`Write Concern </reference/write-concern/>`
         in the {+mdb-server+} manual.

   * - ``hint``
     - | Gets or sets the index to scan for documents. 
         For more information, see the :manual:`hint statement </reference/command/delete/#std-label-deletes-array-hint>`
         in the {+mdb-server+} manual.

   * - ``let``
     - | Specifies a document with a list of values to improve operation readability.
         Values must be constant or closed expressions that don't reference document
         fields. For more information, see the :manual:`let statement
         </reference/command/delete/#std-label-delete-let-syntax>` in the
         {+mdb-server+} manual.

   * - ``comment``
     - | Attaches a comment to the operation. For more information, see the :manual:`delete command
         fields </reference/command/delete/#command-fields>` guide in the
         {+mdb-server+} manual.

> The following code uses the ``delete_many()`` method to delete all documents in

Reviewer: ```suggestion
The following example calls the ``delete_many()`` method to delete all documents in
```

the ``restaurants`` collection that have a ``name`` value containing the string ``"Mongo"``.
It also sets the ``comment`` field of a ``mongocxx::options::delete_options`` instance
to add a comment to the operation:

.. literalinclude:: /includes/write/delete.cpp
   :start-after: start-delete-options
   :end-before: end-delete-options
   :language: cpp
   :dedent:

.. tip::

   If the preceding example used the ``delete_one()`` method instead of
   ``delete_many()``, the driver would delete only the first document that has
   a ``name`` value containing ``"Mongo"``.

Return Value
~~~~~~~~~~~~

The ``delete_one()`` and ``delete_many()`` methods return an instance of
the ``mongocxx::result::delete_result`` class. This class contains the
following member functions:

- ``result()``, which returns the raw bulk write result
- ``deleted_count()``, which returns the number of documents deleted

If the query filter does not match any documents, the driver doesn't delete any
documents and ``deleted_count`` is 0.

> The following example uses the ``delete_many()`` method to delete documents
> that have a ``cuisine`` value of ``"Greek"`` and prints the number of deleted
> documents:

Reviewer: S: Split this to sentence and point out that the `deleted_count()` function is being used
```suggestion
The following example calls the ``delete_many()`` method to delete documents
that have a ``cuisine`` value of ``"Greek"``. It then calls the ``deleted_count()`` member function to print the number of deleted
documents:
```
